### PR TITLE
compiler: remove support for hot-code-reloading

### PR DESCRIPTION
**Pull request by zerbina**
## Summary
HCR was only tested by a single test and made the code-generator logic more complex. The feature is removed in order to reduce the amount of code in `cgen` and to make future changes in the code-generator easier.

The plan is to re-implement support for hot-code-reloading at a point in the future.

## Details
- remove the HCR compiler support
- remove the HCR runtime
- remove the enabled test. The disabled `nimhcr_integration.nim` test is kept
- remove the `nonReloadable` and `executeOnReload` pragmas

---

If wanted, I can also remove the `nimchr_integration.nim` test. The idea behind keeping it was to already have some existing tests when re-implementing HCR support.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

### CODE REVIEW

File: lib/system.nim [hunk 1]
```diff
@@ -529,10 +529,7 @@ proc `..`*[T](b: sink T): HSlice[int, T]
   ##   echo a[.. 2] # @[10, 20, 30]
   result = HSlice[int, T](a: 0, b: b)
 
-when defined(hotCodeReloading):
-  {.pragma: hcrInline, inline.}
-else:
-  {.pragma: hcrInline.}
+{.pragma: hcrInline.}
```
Comment (saem) [inline lib/system.nim]: Could we get rid of this too?

**Comment by saem** [comment]
Nice!

Outside of CI failure I think it's good to go.

File: lib/system.nim (hunk 1 above)
Comment (zerbina) [reply lib/system.nim]: Sure. I only kept it in order to reduce the size of the change-set.

Comment (saem) [reply lib/system.nim]: Makes sense, but more feels is more points. 😁🥳

File: lib/system/cgprocs.nim [hunk 2]
```diff
@@ -13,8 +13,8 @@ type
   LibHandle = pointer       # private type
   ProcAddr = pointer        # library loading and loading of procs:
 
-proc nimLoadLibrary(path: string): LibHandle {.compilerproc, hcrInline, nonReloadable.}
-proc nimUnloadLibrary(lib: LibHandle) {.compilerproc, hcrInline, nonReloadable.}
-proc nimGetProcAddr(lib: LibHandle, name: cstring): ProcAddr {.compilerproc, hcrInline, nonReloadable.}
+proc nimLoadLibrary(path: string): LibHandle {.compilerproc, nonReloadable.}
```
Comment (saem) [inline lib/system/cgprocs.nim]: Personally, I would remove `nonreloadable`, as well.

I know it increases the size, but it more importantly highlights how invasive this whole feature was in the first place.

Comment (saem) [reply lib/system/cgprocs.nim]: this is non-blocking, I'm good to merge it.

Comment (zerbina) [reply lib/system/cgprocs.nim]: It can go too - the required changes are of an okay amount.

Comment (saem) [reply lib/system/cgprocs.nim]: Sweet!

**Comment by zerbina** [comment]
The `.nonReloadable` pragma is now also removed. I missed removing the `executeOnReload` pragma earlier, but that's gone now too.

bors r+

**Comment by saem** [comment]
Yaaaaaasssssssssss!

**Pull request merged**